### PR TITLE
[2.3] Database Media Storage - Design Config Save functions to be Database Media Storage aware

### PR DESCRIPTION
**Pull request by gwharton**
### Description (*)
The functions that process the uploads of the Favicon, Header Image and Transactional Email Logo files do not process the file correctly when in Database Media Storage Mode. This PR fixes that.

### Fixed Issues (if relevant)
1. magento/magento2#21672: Database Media Storage - Design Config fails to save transactional email logo correctly

### Manual testing scenarios (*)
1. Setup Magento
2. Change media storage mode to database
3. Synchronize
4. Save Settings
5. Content->Configuration->Edit->Transactional Emails
6. Upload New Image (but don't save)
7. Verify temporary email logo exists in pub/media
8. Verify temporary email logo exists in database
9. Save Config
10. Verify email logo exists in pub/media
11. Verify email logo exists in database

Repeat for favicon and header logo, ensuring that after save, the file exists correctly in the database.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @gwharton. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by orlangur** [comment]
@aleron75 any update on this?

**Review by aleron75 — APPROVED** [review]

**Comment by VasylShvorak** [comment]
Hi @gwharton could you resolve merge conflicts?

**Comment by gwharton** [comment]
OK, sometimes github does things that I really don't understand.

So i used the github "Resolve conflict" feature to resolve the conflicts. This appeared to work ok, but if you look at the commit history, the resolve conflicts also merged all changes from 2.3-develop into my PR since the PR was created. This means the commit for this PR is now massive!!!! Arghh.

Will I really be re-committing all of the changes made in the last month in 2.3-develop back in on itself????? I don't know what to do to get out of this.

**Comment by aleron75** [comment]
> If you look at the commit history, the resolve conflicts also merged all changes from 2.3-develop into my PR since the PR was created. 

It means that, since you forked the repo, some changes were merged in the main branch and it's normal that your PR now includes those changes because your fork was not up to date anymore.

**Comment by gwharton** [comment]
Ok. Thats great. For one moment i thought my pr would be "recommitting" all those changes twice. The "files changed" section looks correct.

**Review by orlangur — CHANGES_REQUESTED** [review]
Please squash changes into a single commit so that we have perfectly clean history 😉

**Comment by gwharton** [comment]
@orlangur, perhaps I am doing it wrong, I did try and squash them last night into a single commit, but my squashed commit ended up containing all of the changes from the merge of 2.3-develop back into my branch, which I am sure is not what is wanted.

What is the correct way of doing it, when one of the commits, is a merge of 2.3-develop into my branch.

On previous attempts I would have
```
git checkout branch 
git reset --soft HEAD~XXX
git commit
git push --force origin branch
```

**Comment by orlangur** [comment]
@gwharton create a new branch from a fresh `2.3-develop` and then `git merge --squash` from the old branch.

File: app/code/Magento/Theme/Model/Design/Backend/File.php [hunk 1]
```diff
@@ -62,6 +69,7 @@ public function __construct(
         RequestDataInterface $requestData,
         Filesystem $filesystem,
         UrlInterface $urlBuilder,
+        Database $databaseHelper,
```
Comment (sivaschenko) [inline app/code/Magento/Theme/Model/Design/Backend/File.php]: This class is not API, however, let's still try to make this change as backward compatible as possible: could you please add new constructor parameter in a backward compatible way (last optional parameter, property is initialized with object manager if the parameter is not passed)

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @gwharton , thanks for the pull request, please see my suggestion in the review comment

File: app/code/Magento/Theme/Model/Design/Backend/File.php (hunk 1 above)
Comment (gwharton) [reply app/code/Magento/Theme/Model/Design/Backend/File.php]: Done as requested.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4802](https://jira.corp.magento.com/browse/ENGCOM-4802) has been created to process this Pull Request

**Comment by gwharton** [comment]
Backport PR at #21676 has been updated inline with this parent PR, and is ready to go once this has been processed.

**Comment by gwharton** [comment]
Not sure whats going on with the travis build. Not sure its anything to do with me. Looks to be failing attempting to fork a process under php 7.2. I've tried rerunning the travis build twice and it fails with same error. Suspect out of memory during tests.

**Comment by sdzhepa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**